EToCalculator.ts

```diff
@@ -0,0 +1,62 @@
+import { EToData } from "./types";
+
+// The implementation of this algorithm was guided by a step-by-step breakdown (http://edis.ifas.ufl.edu/pdffiles/ae/ae45900.pdf)
+/**
+ * Calculates the reference evapotranspiration using the Penman-Monteith (FAO-56) method (http://www.fao.org/3/X0490E/x0490e07.htm).
+ *
+ * @param etoData The data to calculate the ETo with.
+ * @param elevation The elevation above sea level of the watering site (in meters).
+ * @return The reference evapotranspiration (in millimeters per day).
+ */
+export function calculateETo( etoData: EToData, elevation: number ): number {
+	const avgTemp = ( etoData.maxTemp + etoData.minTemp ) / 2;
+
+	// Adjust the wind speed to a 2m height.
+	const windSpeed = etoData.windSpeed * 4.87 / Math.log( 67.8 * etoData.windSpeedMeasurementHeight - 5.42 );
+
+	const saturationVaporPressureCurveSlope = 4098 * 0.6108 * Math.exp( 17.27 * avgTemp / ( avgTemp + 237.3 ) ) / Math.pow( avgTemp + 237.3, 2 );
+
+	const pressure = 101.3 * Math.pow( ( 293 - 0.0065 * elevation ) / 293, 2.36 );
+
+	const psychrometricConstant = 0.000665 * pressure;
+
+	const deltaTerm = saturationVaporPressureCurveSlope / ( saturationVaporPressureCurveSlope + psychrometricConstant * ( 1 + 0.34 * windSpeed ) );
+
+	const psiTerm = psychrometricConstant / ( saturationVaporPressureCurveSlope + psychrometricConstant * ( 1 + 0.34 * windSpeed ) );
+
+	const tempTerm = ( 900 / ( avgTemp + 273 ) ) * windSpeed;
+
+	const minSaturationVaporPressure = 0.6108 * Math.exp( 17.27 * etoData.minTemp / ( etoData.minTemp + 237.3 ) );
+
+	const maxSaturationVaporPressure = 0.6108 * Math.exp( 17.27 * etoData.maxTemp / ( etoData.maxTemp + 237.3 ) );
+
+	const avgSaturationVaporPressure = ( minSaturationVaporPressure + maxSaturationVaporPressure ) / 2;
+
+	const actualVaporPressure = ( minSaturationVaporPressure * etoData.maxHumidity / 100 + maxSaturationVaporPressure * etoData.minHumidity / 100 ) / 2;
+
+	const inverseRelativeEarthSunDistance = 1 + 0.033 * Math.cos( 2 * Math.PI / 365 * etoData.dayOfYear );
```

Comment:
If we add a `timestamp` field to our `WateringData` structure then you can calculate dayOfYear from that rather than having to pass in. Feels like a more generic solution.

Reply:
We discussed adding timestamps to WateringData/EToData/WeatherData objects before but I forget if we established what they should be set to. Should it be the time at the start of the 24 hour period the data is based on?

Reply:
I think the `timestamp` field on the response from `get...Data` should mark the time at which the api call was made. It marks the "when" of the contained data i.e. the `temp` as at `timestamp` or the forecast for tomorrow as at `timestamp`. This is a bit like the `weatherProvider` field that tells us the "where" the fields come from. Again, not a big point as ETo calculation could just derive the `dayOfYear` from the current time.

Reply:
I'd like to avoid always using the current date for `dayOfTheYear` so that we can use the function to calculate the ETo for other days in case a need for that ever arises

I think that the timestamp should track the time that the generated data is for. Any function that wants to know what time the API call was made can just note the time when it called `getWateringData`, but there isn't currently a way to know what time period the data actually represents (for different combinations of data types and WeatherProviders it could either be the previous calendar day, the previous 24 hours, or a 24 hour time period starting at the current time).

Reply:
Concur that the timestamp should track the time period that the generated data represents (or is targeted for).  That allows further generalization of this collection of processes.

Reply:
So a few different thoughts going on here and a couple of other thread, so will try to pull together:
1) `timestamp`: I meant to represent when the data was captured from the api and therefore the point in time that the data was valid i.e. `wateringData.temp` was the average temp at that particular point in time. This is a standard data model hygiene factor that says the source should timestamp data rather than the downstream destinations otherwise you end up with differences across Weather Service, Firmware or App. As I say just a hygiene factor.
2) `dayOfYear`: at the moment the applicable period can be derived within `calculateEto()` from `now()` so it is not a necessary field to pass in via `ETOData`
3) Future-Proof: looking forward, I think `dayOfYear/applicablePeriod` should be a parameter to `calculateETO` in order to allow adhoc calculation e.g. for `getWeatherData` to call `calculateETO` for each day of the forecast so we can provide the end-user with an outlook of ETo over the week

Does that make sense ?